### DISCUSSION TITLE
Stream doesn't get SHUTDOWN_COMPLETE

### DISCUSSION THREAD

**Discussion by ManickaP** [microsoft/msquic#2047] [Q&A] [answered]
Client finished sending and gets SEND_SHUTDOWN_COMPLETE, then it calls ABORT_READ, and GRACEFUL StreamShutdown, yet it never receives SHUTDOWN_COMPLETE (until the connection gets closed minutes later).

On the server side, the stream shutdowns properly (receiving SHUTDOWN_COMPLETE) and destroys itself.

Attaching [msquic.log](https://github.com/microsoft/msquic/files/7288426/msquic.log) extract.

And here's the sequence of shutdown calls from the client logs:
```
21:05:04.4342388[Info] thisOrContextObject: State#21233793, memberName: StartShutdown, message: [strm][0x7F37440A3120] Stream StartShutdown flags=ABORT_RECEIVE error=268
21:05:04.4342554[Info] thisOrContextObject: State#21233793, memberName: StartShutdown, message: [strm][0x7F37440A3120] Stream StartShutdown flags=GRACEFUL error=0
21:05:04.4342595[Info] thisOrContextObject: State#21233793, memberName: StartShutdown, message: [strm][0x7F37440A3120] Stream StartShutdown flags=ABORT_RECEIVE error=4294967295
```

Are we doing something wrong?

**Comment by nibanks** [comment, marked as answer]
Figured out the bug. There is a race that's not properly handled. Opened issue https://github.com/microsoft/msquic/issues/2048.
